src/systems/camera-system.js

```diff
@@ -0,0 +1,182 @@
+    const l1 = Math.abs((halfYExtents * margin) / Math.tan(halfVertFOV));
+    const l2 = Math.abs((halfXExtents * margin) / Math.tan(halfHorFOV));
+    const l3 = Math.abs(box.max.z - center.z) + Math.max(l1, l2);
+    const l = object.el.sceneEl.is("vr-mode") ? Math.max(0.25, l3) : l3;
+    return l;
+  };
+})();
+
+const moveRigSoCameraLooksAtObject = (function() {
+  const cqInv = new THREE.Quaternion();
+  const owq = new THREE.Quaternion();
+  const v1 = new THREE.Vector3();
+  const v2 = new THREE.Vector3();
+  const owp = new THREE.Vector3();
+
+  return function moveRigSoCameraLooksAtObject(rig, camera, object) {
+    object.getWorldQuaternion(owq);
+    object.getWorldPosition(owp);
+    cqInv.copy(camera.quaternion).inverse();
+    rig.quaternion.copy(owq).multiply(cqInv);
+
+    v1.set(0, 0, 1)
+      .multiplyScalar(calculateViewingDistance(camera, object))
+      .applyQuaternion(owq);
+    v2.copy(camera.position).applyQuaternion(rig.quaternion);
+    rig.position.subVectors(v1, v2).add(owp);
+
+    rig.matrixNeedsUpdate = true;
+  };
+})();
+
+let numCameraModes = 0;
+export const CAMERA_MODE_FIRST_PERSON = numCameraModes++;
+export const CAMERA_MODE_THIRD_PERSON_NEAR = numCameraModes++;
+export const CAMERA_MODE_THIRD_PERSON_FAR = numCameraModes++;
+export const CAMERA_MODE_INSPECT = numCameraModes++;
+
+const m2 = new THREE.Matrix4();
+const m3 = new THREE.Matrix4();
+const offset = new THREE.Vector3();
+export class CameraSystem {
+  constructor() {
+    this.mode = CAMERA_MODE_FIRST_PERSON;
+    waitForDOMContentLoaded().then(() => {
+      this.playerCamera = document.getElementById("player-camera");
+      this.playerRig = document.getElementById("player-rig");
+      this.cameraEl = document.getElementById("experimental-camera");
+      this.rigEl = document.getElementById("experimental-rig");
+      this.rigEl.object3D.matrixWorldNeedsUpdate = true;
+      this.rigEl.object3D.matrixIsModified = true;
+    });
+  }
+  nextMode() {
+    if (!qsTruthy("thirdPerson")) return;
+
+    if (this.mode === CAMERA_MODE_INSPECT) {
+      this.uninspect();
+      return;
+    }
+
+    this.mode = (this.mode + 1) % numCameraModes;
+    if (this.mode === CAMERA_MODE_FIRST_PERSON) {
+      AFRAME.scenes[0].renderer.vr.setPoseTarget(this.playerCamera.object3D);
+    } else if (this.mode === CAMERA_MODE_THIRD_PERSON_NEAR || this.mode === CAMERA_MODE_THIRD_PERSON_FAR) {
+      AFRAME.scenes[0].renderer.vr.setPoseTarget(this.cameraEl.object3D);
+    } else if (this.mode === CAMERA_MODE_INSPECT) {
+      this.mode = CAMERA_MODE_FIRST_PERSON; // skip inspect for now
+    }
+  }
+  inspect(o) {
+    if (this.mode !== CAMERA_MODE_INSPECT) {
+      this.preInspectMode = this.mode;
+    }
+    this.mode = CAMERA_MODE_INSPECT;
+    this.inspected = o;
+    moveRigSoCameraLooksAtObject(this.rigEl.object3D, this.cameraEl.object3D, this.inspected);
+  }
+  uninspect() {
+    this.inspected = null;
+    if (this.mode !== CAMERA_MODE_INSPECT) {
```

Comment:
nit: can make one liner `if (this.mode !== CAMERA_MODE_INSPECT) return`, also probably would put as first line of the function to show it's a guard clause

Reply:
I still want `this.inspected = null` to run in the case this is called from `nextMode` but I did make the check a one liner.